magma/backend/mlir/hardware_module.py

```diff
@@ -362,33 +362,61 @@ def make_concat(self, operands, result):
 
         hw.ArrayConcatOp(operands=new_operands, results=[result])
 
+    def _make_mem_reg(self, name, N, T):
+        mem_type = hw.InOutType(hw.ArrayType((N,), T))
+        mem = self._ctx.new_value(mem_type)
+        sv.RegOp(name=name, results=[mem])
+        return mem
+
+    def _make_mem_read(self, target, value, has_enable, name_suffix=""):
+        if not has_enable:
+            sv.ReadInOutOp(operands=[target], results=[value])
+            return None, None
+        read_temp = self._ctx.new_value(target.type.T)
+        sv.ReadInOutOp(operands=[target], results=[read_temp])
+        read_reg = self._ctx.new_value(target.type)
+        sv.RegOp(name=f"read_reg{name_suffix}", results=[read_reg])
```

Comment:
these helper functions probably shouldn't decide name, so can have `name_suffix` just be the whole `name` instead